scalameta/quasiquotes/shared/src/main/scala-3/scala/meta/internal/quasiquotes/ReificationMacros.scala

```diff
@@ -0,0 +1,688 @@
+  }
+
+  private def instantiateDialect(mode: Mode): (Dialect, Expr[Dialect]) = {
+    val dialectExpr = Expr.summon[scala.meta.Dialect]
+    // For now only a predefined set of dialects is supported.
+    // See comment in the scala 2 counterpart file.
+    val underlyingDialect = {
+      def instantiateStandardDialect(sym: Symbol): Option[Dialect] = {
+        val dialectsSym = Symbol.classSymbol("scala.meta.dialects.package$").companionModule
+        if (dialectsSym == sym.owner.companionModule)
+          if (dialectsSym.methodMember(sym.name).isEmpty) None else Dialect.standards.get(sym.name)
+        else None
+      }
+
+      // allow `scala.meta.dialects.Scala211`
+      val standardDialectReference = dialectExpr
+        .flatMap(expr => instantiateStandardDialect(expr.asTerm.tpe.termSymbol))
+
+      // allow `scala.meta.dialects.current`
+      def standardDialectSingleton = dialectExpr.flatMap {
+        _.asTerm.tpe match
+          case termRef @ TermRef(prefix, _) => // TermRef(scala.meta.Dialect, method current)
+            instantiateStandardDialect(prefix.memberType(termRef.termSymbol).termSymbol)
+          case _ => None
+      }
+
+      standardDialectReference.orElse(standardDialectSingleton).getOrElse {
+        val suggestion =
+          s"to fix this, import something from scala.meta.dialects, e.g. scala.meta.dialects.${Dialect.current}"
+        val message =
+          s"${dialectExpr.get.show} of type ${dialectExpr.get.asTerm.tpe.show} is not supported by quasiquotes ($suggestion)"
+        report.errorAndAbort(message)
+      }
+    }
+    (
+      if (mode.isTerm) underlyingDialect.unquoteTerm(mode.multiline)
+      else underlyingDialect.unquotePat(mode.multiline),
+      dialectExpr.get
+    )
+  }
+  private def instantiateParser(qType: QuasiquoteType): MetaParser = {
+    val parserModule = Symbol.classSymbol(qType.parserClass())
+    val parsersModuleClass = Class
+      .forName("scala.meta.quasiquotes.package$", true, this.getClass.getClassLoader)
+    val parsersModule = parsersModuleClass.getField("MODULE$").get(null)
+    val parserModuleGetter = parsersModule.getClass.getDeclaredMethod(parserModule.name.toString)
+    val parserModuleInstance = parserModuleGetter.invoke(parsersModule)
+    val parserMethod = parserModuleInstance.getClass.getDeclaredMethods.find(_.getName == "parse")
+      .head
+    (input: Input, dialect: Dialect) =>
+      try parserMethod.invoke(parserModuleInstance, input, dialect).asInstanceOf[MetaTree]
+      catch { case ex: java.lang.reflect.InvocationTargetException => throw ex.getTargetException }
+  }
+  private def parseSkeleton(parser: MetaParser, input: Input, dialect: Dialect): MetaTree =
+    try parser(input, dialect)
+    catch {
+      case TokenizeException(pos, message) => report.errorAndAbort(message)
+      case ParseException(pos, message) => report.errorAndAbort(message)
```

Comment:
`pos` is not used. is it possible to avoid unapply then and just do
```
  case ex @ (_: TokenizeException | _: ParseException) => report.errorAndAbort(ex.getMessage)
```

Reply:
I ended up changing this so that we do use the position now (forgot to do that before)